src/main/java/org/candlepin/model/Product.java

```diff
@@ -1308,42 +1309,46 @@ public int getEntityVersion() {
             // again, doesn't implement .hashCode reliably on the proxy collections. So, we have to
             // manually step through these and add the elements to ensure the hash code is
             // generated properly.
-            int accumulator;
+            // Additionally, the algorithm used by HashCodeBuilder is order dependent, so we must be
+            // sure to add our children entities into it in a consistent manner.
 
             if (!this.providedProducts.isEmpty()) {
-                accumulator = 0;
-                for (Product product : this.providedProducts) {
-                    accumulator += (product != null ? product.getEntityVersion() : 0);
-                }
+                builder.append("provided_products");
 
-                builder.append(accumulator);
```

Comment:
I don't understand how this had an order dependency problem. If we had let's say 2 provided products, one with entity version 20 and another with entity version 30, the accumulator would always end up being 50, regardless of the order of those 2 providedProducts being processed?

Reply:
The ordering thing was an issue the accumulator was intended to solve, and did, but also introduced two issues: if two different products had collections of children whose accumulated entity version or hashcode just happened to be the same, it would cause them to collide -- even if they occurred in different sets of children (i.e. product 1 has branding with an accumulated hashcode of A, and product 2 with content with an accumulated version of A, they will resolve to the same version).

New tests have been added to cover both cases, and may be a better way of understanding what I'm struggling to explain here.

Reply:
I understand now, thanks